File: src/main/java/com/baulsupp/oksocial/authenticator/oauth2/Oauth2ServiceDefinition.java
package com.baulsupp.oksocial.authenticator.oauth2;

import com.baulsupp.oksocial.AbstractServiceDefinition;

public class Oauth2ServiceDefinition extends AbstractServiceDefinition<Oauth2Token> {
  public Oauth2ServiceDefinition(String apiHost, String serviceName, String shortName,
      String apiDocs) {
    super(apiHost, serviceName, shortName, apiDocs);
  }

  public Oauth2Token parseCredentialsString(String s) {
    String[] parts = s.split(":");

    if (parts.length < 4) {
      return new Oauth2Token(parts[0]);
    } else {
      return new Oauth2Token(parts[0], parts[1], parts[2], parts[3]);
    }
  }

  public String formatCredentialsString(Oauth2Token credentials) {
    if (credentials.refreshToken.isPresent()
        && credentials.clientId.isPresent()
        && credentials.clientSecret.isPresent()) {
      return credentials.accessToken
          + ":"
          + credentials.refreshToken.get()
          + ":"
          + credentials.clientId.get()
          + ":"
          + credentials.clientSecret.get();
    } else {
      return credentials.accessToken;
    }
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/authenticator/oauth2/Oauth2ServiceDefinition.kt
package com.baulsupp.oksocial.authenticator.oauth2

import com.baulsupp.oksocial.AbstractServiceDefinition

class Oauth2ServiceDefinition(apiHost: String, serviceName: String, shortName: String,
                              apiDocs: String, accountsLink: String?) : AbstractServiceDefinition<Oauth2Token>(apiHost, serviceName, shortName, apiDocs, accountsLink) {

    override fun parseCredentialsString(s: String): Oauth2Token {
        val parts = s.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        return if (parts.size < 4) {
            Oauth2Token(parts[0])
        } else {
            Oauth2Token(parts[0], parts[1], parts[2], parts[3])
        }
    }

    override fun formatCredentialsString(credentials: Oauth2Token): String {
        if (credentials.refreshToken != null
                && credentials.clientId != null
                && credentials.clientSecret != null) {
            return "${credentials.accessToken}:${credentials.refreshToken}:${credentials.clientId}:${credentials.clientSecret}"
        } else {
            return credentials.accessToken
        }
    }
}
